do-not-commit/default-mutable-dict.py

```diff
@@ -0,0 +1,393 @@
+import copy
+
+# ruleid: default-mutable-dict
+def assign_func1(default={}):
+    default["potato"] = 5
+
+
+# ruleid: default-mutable-dict
+def assign_func2(default={}):
+    for x in range(10):
+        default[x] = 1
+
+
+# todoruleid: default-mutable-dict
+def assign_func3(default={}):
+    x = default
+    x[3] = 2
+
+
+# ruleid: default-mutable-dict
+def assign_func4(x=1, default={}):
+    default["1"] = 1
+
+
+# ruleid: default-mutable-dict
+def assign_func5(default={}):
+    if not default:
+        default["1"] = "test"
+
+
+# ruleid: default-mutable-dict
+def assign_func6(default={}, x="string"):
+    default[1] = 0
+
+
+# todoruleid: default-mutable-dict
+def assign_func7(default={}):
+    if True:
+        default = dict(default)
+    else:
+        default[1] = 21
+
+
+# ruleid: default-mutable-dict
+def assign_func8(default={}):
+    while True:
+        default[1] = 4
+        break
+
+
+# ruleid: default-mutable-dict
+def update_func1(default={}):
+    default.update({1: 2})
+
+
+# ruleid: default-mutable-dict
+def update_func2(default={}):
+    for x in range(10):
+        default.update({1: 2})
+
+
+# todoruleid: default-mutable-dict
+def update_func3(default={}):
+    x = default
+    x.update({1: 2})
+
+
+# ruleid: default-mutable-dict
+def update_func4(x=1, default={}):
+    default.update({1: 2})
+
+
+# ruleid: default-mutable-dict
+def update_func5(default={}):
+    if not default:
+        default.update({1: 2})
+
+
+# ruleid: default-mutable-dict
+def update_func6(default={}, x="string"):
+    default.update({1: 2})
+
+
+# todoruleid: default-mutable-dict
+def update_func7(default={}):
+    if True:
+        default = dict(default)
+    else:
+        default.update({1: 2})
+
+
+# ruleid: default-mutable-dict
+def update_func8(default={}):
+    while True:
+        default.update({1: 2})
+        break
+
+
+# ruleid: default-mutable-dict
+def setdefault_func1(default={}):
```

Comment:
Function setdefault_func1 mutates default dict default. Instances of default function arguments in Python are shared between function calls.  Mutating a default function argument changes the result for all function consumers. This will most likely cause unexpected results, or lead to security vulnerabilities whereby one function consumer can either view or modify the data of another function consumer. ([.common-mistakes.default-mutable-dict.default-mutable-dict](https://semgrep.dev/r/.common-mistakes.default-mutable-dict.default-mutable-dict))
#
<div>
<a href="https://semgrep.dev">
<img align="right" src="https://r2c.dev/semgrep_icon_text.svg" height=28px alt="" />
</a>

<sub> [How do I suppress this?](https://semgrep.dev/docs/running-rules/#managing-findings) **·** [Report a bug](https://github.com/returntocorp/semgrep-rules/issues/new?assignees=&labels=&template=rule_bug_report.md&title=.common-mistakes.default-mutable-dict.default-mutable-dict) **·** [Manage rule](https://semgrep.dev/manage/projects)</sub>
</div>